### ISSUE TITLE
[Bug] Hardnet do not pass gradcheck

### ISSUE THREAD

**Issue by edgarriba** [kornia/kornia#503]
Recently introduced hardnet descriptor (https://github.com/kornia/kornia/pull/498) is not able to pass `gradcheck` tests.

The test is disabled right now and might need some investigation.
Link to the failing tests: https://github.com/kornia/kornia/blob/ee668e38ee8bea32c092538a73add1b1d8f550e2/test/feature/test_hardnet.py#L26-L32

Raised error: https://pastebin.com/pKVezb1d

**Comment by DagnyT** [comment]
I've noticed several things:

1. doesn't work with dropout. 
2. torch.backends.cudnn.deterministic = True is needed here
3. the problem stays with _normalize_input method, I assume it might violate this "If any checked tensor in input has overlapping memory, i.e., different indices pointing to the same memory address (e.g., from torch.expand()), this check will likely fail because the numerical gradients computed by point perturbation at such indices will change values at all other indices that share the same memory address.", but not sure.
Otherwise, without dropout and normalization outside the model, it passes test with gradcheck default params.

**Comment by edgarriba** [comment]
@DagnyT thanks for checking. This are my findings:

![Screenshot from 2020-04-01 20-51-54](https://user-images.githubusercontent.com/5157099/78175840-d1dd2b00-745b-11ea-8a63-fb9ccbf6acc0.png)

Now the question is why detach was needed and how do we test this without dropout ?  /cc @ducha-aiki

**Comment by DagnyT** [comment]
dropout - there might be a beautiful way, hacky way - 
hardnet = HardNet()
hardnet.features[18].p=0

detach() is a problem; there was instability during training without it.

**Comment by edgarriba** [comment]
Great! Dropout solved.

Have you guys checked this instability with new pytorch versions?
